### PR TITLE
Create deployment automation

### PR DESCRIPTION
**Pull request by cabutlermit**
### Why these changes are being introduced

This is the theme deployment automation for Omeka S.

### How this addresses that need

* Create stage and prod workflows for GitHub Actions to automatically deploy the theme to the Omeka S servers
* Create documentation (README and reference doc, ADR#0004)

### Side effects of this change

Once merged to main, the Actions will be enabled for commits and release on `main`.

### Other Relevant Information

At this point, the self-hosted runner for stage is configured for this repository. But, the self-hosted runner for prod has not yet been enabled. So, the prod GitHub Workflow will fail until we create that runner and configure it on the prod Omeka server.

### Relevant ticket(s)

* https://mitlibraries.atlassian.net/browse/POST-60

### CODE REVIEW

**Comment by cabutlermit** [comment]
@zotoMIT I'm tagging you to make sure that I didn't miss anything obvious with the self-hosted runner setup and use in the workflows...

File: .github/workflows/prod-deploy.yml
Comment (zotoMIT) [inline .github/workflows/prod-deploy.yml:1]: I think this is named wrong?

**Review by zotoMIT — COMMENTED** [review]
Other than i think a mistake in the naming this LGTM

File: .github/workflows/prod-deploy.yml
Comment (cabutlermit) [reply .github/workflows/prod-deploy.yml:1]: oops!

**Review by zotoMIT — APPROVED** [review]
LGTM

**Review by matt-bernhardt — COMMENTED** [review]
I've got a lot of questions to confirm my understanding, but only one of them might rise to the level of requesting changes:

The way this is structured, it would prevent us from using the staging instance to get stakeholder approval during feature development. Without a dev instance, that would leave us without a great option for those conversations:

- Do code review before stakeholder approval (would likely result in more PRs being needed, as the bulk of changes currently happen during talks with stakeholders rather than code review)
- Deploy work-in-progress to the sandbox instance (not ideal as the sandbox is a sub-directory install and doesn't currently work with all the features of this theme, for example)
- Ask stakeholders to run, and synchronize content in, a local Lando container (feels like too high a bar to ask of colleagues)
- Get stakeholder approval via screenshots or other indirect experience (Zoom pairings or demonstrations, etc)

If deploys to staging only after PRs merge is the only way to handle this, then we can try to identify a workflow that satisfies everyone - but I'd rather enable additional deploys to staging.

File: Makefile
Comment (cabutlermit) [reply Makefile:30]: oops! Will be fixed in next push.

File: Makefile
Comment (cabutlermit) [reply Makefile:31]: oops! Will be fixed in next push.

File: .github/workflows/stage-deploy.yml
Comment (cabutlermit) [reply .github/workflows/stage-deploy.yml:4]: Yes, that is a good point, with a caveat. The "deploy-to-prod" is triggered by a published release. So, pre-releases will not trigger the production deploy nor will just regular tags not marked as releases.

I guess my question back to you would be the following: Would you ever create a "release" that does not include changes to the theme? You can merge to main as much as you want (which, if the changes to the repo occur in the designated folders, will trigger a deploy to the staging server), but nothing will deploy to production without an official tagged release. I was working under the assumption that tooling updates could be commited to the main branch, but that these would not warrant a release. maybe my assumption is incorrect? or maybe we need to more clearly document how/when we are using releaes?

File: .github/workflows/prod-deploy.yml
Comment (cabutlermit) [reply .github/workflows/prod-deploy.yml:4]: No, only releases; not pre-releases.  See [release](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#release) in the GHA documentation.

File: docs/architecture-decisions/0004-github-runner-for-automated-deployment.md
Comment (cabutlermit) [reply docs/architecture-decisions/0004-github-runner-for-automated-deployment.md:1]: ha ha! You caught me in some undercooked copy-pasta... will fix shortly.

File: .github/CODEOWNERS
Comment (cabutlermit) [reply .github/CODEOWNERS:15]: This is fixed -- the InfraEng Team had just "read" access to this repo and I just updated that to "write".

**Comment by cabutlermit** [comment]
> I've got a lot of questions to confirm my understanding, but only one of them might rise to the level of requesting changes:
> 
> The way this is structured, it would prevent us from using the staging instance to get stakeholder approval during feature development. Without a dev instance, that would leave us without a great option for those conversations:
> 
>     * Do code review before stakeholder approval (would likely result in more PRs being needed, as the bulk of changes currently happen during talks with stakeholders rather than code review)
> 
>     * Deploy work-in-progress to the sandbox instance (not ideal as the sandbox is a sub-directory install and doesn't currently work with all the features of this theme, for example)
> 
>     * Ask stakeholders to run, and synchronize content in, a local Lando container (feels like too high a bar to ask of colleagues)
> 
>     * Get stakeholder approval via screenshots or other indirect experience (Zoom pairings or demonstrations, etc)
> 
> If deploys to staging only after PRs merge is the only way to handle this, then we can try to identify a workflow that satisfies everyone - but I'd rather enable additional deploys to staging.

Your analysis is correct -- this model is built on the idea that stakeholder review and approval happens before the PR is merged to the `main` branch. Once merged, full review of the changes is evaluated in the stage instance before creating the release on `main`. This model has worked well in some of our other applications, but I can totally understand your concerns with this repo and the lack of widely available dev instance.

We do have some options though.

1. We can enable the `workflow_dispatch` option for a new `dev-deploy.yml` workflow. This will give developers a "button" in the Actions tab of the repo that will allow you to manually trigger a deploy to the stage instance from the most recent commit on a feature branch. This would change the stage instance into more of a stage/dev instance in regard to theme development.
2. We could put some effort into creating a dev instance of Omeka S in our Dev1 AWS Account and then have a proper development instance that would work with all of our automation for exactly this kind of stakeholder review and testing.

Option (1) is the much quicker option and is probably the simplest and quickest path forward in the short term. Option (2) is a bit of work, and the long-term benefit of a standalone dev instance is unclear (it's not clear to me how much additional theme modification will be happening after we go live).

File: .github/workflows/stage-deploy.yml
Comment (cabutlermit) [reply .github/workflows/stage-deploy.yml:10]: See my response to your primary comment below.

File: .github/CODEOWNERS
Comment (cabutlermit) [reply .github/CODEOWNERS:7]: Probably.  I'm still working through some of my thoughts on how we want to make use of the `CODEOWNERS` file to share/track information about the repo. I think that it would be sufficient to just have the `@mitlibraries/engx` here, but keeping `@matt-bernhardt` here helps indicate that you are the primary owner of this repository. I believe (but could maybe be convinced otherwise) that there should always be at least one individual who is flagged as an owner (instead of just a group as an owner) to avoid "well, someone else in my group will deal with this" problem when the primary owner is a group.

**Comment by cabutlermit** [comment]
@matt-bernhardt I think that I addressed most (all?) of your concerns with the recent updates to the branch in this PR, including an Action that can be triggered from the Actions tab manually by the dev to force a theme deploy to the stage server from a feature branch. Can you review (and resolve any comments that you agree are resolved)? and then we can fine-tune for any remaining issues?

**Comment by matt-bernhardt** [comment]
Yeah, sorry for the delay - I meant to get back to you yesterday and then fell down a rabbit hole. I'm writing responses now.

File: .github/CODEOWNERS
Comment (matt-bernhardt) [reply .github/CODEOWNERS:7]: I can see that argument, yeah. Ultimately I think this is something that should be decided across DLS and between team leads like yourself and Jeremy, but in terms of this PR I'm fine with it being this way at this time.

File: .github/workflows/stage-deploy.yml
Comment (matt-bernhardt) [reply .github/workflows/stage-deploy.yml:4]: I think your assumption is valid, reading your logic and thinking through everything more. I've also realized that, by including the `config/` path, we only need bump the theme version number to qualify as something that deploys to staging - so the blind spot I was worried about is even narrower than I thought initially. If there's a tooling change that is significant enough to warrant a standalone deploy to production, it feels appropriate that we'd also bump the theme version, which would then trigger a staging deploy.

TL;DR - all good here.

**Review by matt-bernhardt — APPROVED** [review]
> We do have some options though.
> 
> 1. We can enable the `workflow_dispatch` option for a new `dev-deploy.yml` workflow. This will give developers a "button" in the Actions tab of the repo that will allow you to manually trigger a deploy to the stage instance from the most recent commit on a feature branch. This would change the stage instance into more of a stage/dev instance in regard to theme development.
> 
> 2. We could put some effort into creating a dev instance of Omeka S in our Dev1 AWS Account and then have a proper development instance that would work with all of our automation for exactly this kind of stakeholder review and testing.
> 
> 
> Option (1) is the much quicker option and is probably the simplest and quickest path forward in the short term. Option (2) is a bit of work, and the long-term benefit of a standalone dev instance is unclear (it's not clear to me how much additional theme modification will be happening after we go live).

I agree that option (1) is preferable to (2), and would address my workflow concerns. It looks like the new `dev-manual.yml` accomplishes that?

I'm looking forward to trying this out! Thanks for talking through all this with me.

:shipit:

**Pull request merged**
